appengine/go11x/warmup/warmup.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2018 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// [START gae_go111_warmup]
+
+// Sample warmup demonstrates usage of the /_ah/warmup handler.
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"cloud.google.com/go/storage"
+)
+
+var startupTime time.Time
+var client *storage.Client
+
+func main() {
+	if err := setup(context.Background()); err != nil {
+		log.Fatalf("setup: %v", err)
+	}
+
+	http.HandleFunc("/_ah/warmup", func(w http.ResponseWriter, r *http.Request) {
```

Comment:
I wonder if `setup` should be inline and warmup should be a different function, with a comment for what you might do in a warmup function?

It's not clear that `setup()` is key to understanding warmup. But, feel free to disagree if there was a goal with this setup.

Reply:
My intention was to perform all warmup tasks as part of main and leave the warmup handler dedicated to logging that the official warmup request had successful triggered.

I wanted to demonstrate it this way because it seemed more unit testable to have a separate setup that could receive any context.

Would the concern be addressed if I rename setup to warmup and add comments explaining this?

Reply:
@tbpg could you look at this again? I've made some comment and function name changes which hopefully help to clarify.

Reply:
I think my confusion comes my gut saying, "we should do warmup stuff during the warmup request." But, maybe the warmup request is a signal the instance is started and ready to go? In which case, I think naming the function `setup` is better than `warmup()` (otherwise, why is `warmup()` not connected to `/_ah/warmup`?). Whatever you think is most clear.

Reply:
Switching to setup but keeping the comments.

The reason is that a warmup request is determined by scheduler magic, and if I understand correctly, doesn't preclude cold starts. It also ensures that local or other hosting of the application that uses a "normal" startup has the correct initialization.

I can't think of any initialization steps that I would **only** want on warmup but not on a cold start.